adam_core/dynamics/propagation.py

```diff
@@ -0,0 +1,174 @@
+        Orbit propagated to each MJD with position in units of au and velocity in units
+        of au per day.
+    """
+    r = orbit[0:3]
+    v = orbit[3:6]
+    dt = t1 - t0
+
+    lagrange_coeffs, stumpff_coeffs, chi = calc_lagrange_coefficients(
+        r, v, dt, mu=mu, max_iter=max_iter, tol=tol
+    )
+    r_new, v_new = apply_lagrange_coefficients(r, v, *lagrange_coeffs)
+
+    return jnp.array([r_new[0], r_new[1], r_new[2], v_new[0], v_new[1], v_new[2]])
+
+
+# Vectorization Map: _propagate_2body
+_propagate_2body_vmap = jit(
+    vmap(_propagate_2body, in_axes=(0, 0, 0, None, None, None), out_axes=(0))
+)
+
+
+def propagate_2body(
+    orbits: Orbits,
+    times: Time,
+    mu: float = MU,
+    max_iter: int = 1000,
+    tol: float = 1e-14,
+) -> Orbits:
+    """
+    Propagate orbits using the 2-body universal anomaly formalism.
+
+    Parameters
+    ----------
+    orbits : `~jax.numpy.ndarray` (N, 6)
+        Cartesian orbits with position in units of au and velocity in units of au per day.
+    times : `~astropy.time.core.Time` (M)
+        Epochs to which to propagate each orbit. If a single epoch is given, all orbits are propagated to this
+        epoch. If multiple epochs are given, then each orbit to will be propagated to each epoch.
+    mu : float, optional
+        Gravitational parameter (GM) of the attracting body in units of
+        au**3 / d**2.
+    max_iter : int, optional
+        Maximum number of iterations over which to converge. If number of iterations is
+        exceeded, will return the value of the universal anomaly at the last iteration.
+    tol : float, optional
+        Numerical tolerance to which to compute universal anomaly using the Newtown-Raphson
+        method.
+
+    Returns
+    -------
+    orbits : `~adam_core.orbits.orbits.Orbits` (N*M)
+        Orbits propagated to each MJD.
+    """
+    # Lets extract the cartesian orbits and times from the orbits object
+    cartesian_orbits = orbits.coordinates.values
+    t0 = orbits.coordinates.time.to_astropy().tdb.mjd
+    t1 = times.tdb.mjd
+    orbit_ids = orbits.orbit_id.to_numpy(zero_copy_only=False)
+    object_ids = orbits.object_id.to_numpy(zero_copy_only=False)
+
+    # Lets stack the orbits into a single array shaped by the number of orbits and number of times
+    # and then pass this to the vectorized map version of _propagate_2body
+    n_orbits = cartesian_orbits.shape[0]
+    n_times = len(times)
+    orbit_ids_ = np.repeat(orbit_ids, n_times)
+    object_ids_ = np.repeat(object_ids, n_times)
+    orbits_array_ = np.repeat(cartesian_orbits, n_times, axis=0)
+    t0_ = np.repeat(t0, n_times)
+    t1_ = np.tile(t1, n_orbits)
+
+    orbits_propagated = _propagate_2body_vmap(
+        orbits_array_, t0_, t1_, mu, max_iter, tol
+    )
+    orbits_propagated = np.array(orbits_propagated)
+
+    cartesian_covariances = orbits.coordinates.covariance.to_matrix()
+    if not np.all(np.isnan(cartesian_covariances)):
```

Comment:
Trying to work this out - this is "if any of the cells in any of the covariance matrices is non-null", right?

Is `transform_covariances_jacobian` OK with _some_ nulls in a particular covariance matrix?

Reply:
I haven't tried but I suspect it will fail. We may need to approximate nulls in a covariance matrix as some very large number to give them the appearance of being unconstrained. In most cases, I suspect if we have a full state vector we will have either no covariance matrix or a fully defined one.

Reply:
To clarify, if any cells in this 3D matrix of (N, 6, 6) values are null this will return False and not execute the propagated covariance.

Reply:
So theoretically, we could check along the first axis and then do some masking to pass only those covariances that are fully defined.

Reply:
> To clarify, if any cells in this 3D matrix of (N, 6, 6) values are null this will return False and not execute the propagated covariance.

Are you sure? I think that's backwards - this is "not all nan", but it sounds like you want "all not nan." (`np.all(~np.isnan(cartesian_covariances))`)

```
>>> import numpy as np
>>> valid = np.ones((6, 6), np.float64())
>>> partial_valid = np.ones((6, 6), np.float64())
>>> partial_valid[0][0] = float('nan')
>>> data = np.array([valid, partial_valid])
>>> print(data)
[[[ 1.  1.  1.  1.  1.  1.]
  [ 1.  1.  1.  1.  1.  1.]
  [ 1.  1.  1.  1.  1.  1.]
  [ 1.  1.  1.  1.  1.  1.]
  [ 1.  1.  1.  1.  1.  1.]
  [ 1.  1.  1.  1.  1.  1.]]

 [[nan  1.  1.  1.  1.  1.]
  [ 1.  1.  1.  1.  1.  1.]
  [ 1.  1.  1.  1.  1.  1.]
  [ 1.  1.  1.  1.  1.  1.]
  [ 1.  1.  1.  1.  1.  1.]
  [ 1.  1.  1.  1.  1.  1.]]]
>>> not np.all(np.isnan(data))
True

Reply:
Oh s***. You are completely correct. It's because this used to be a masked array and masks are True when they are masked so it's inverted here.. I'll fix it in the 2-body ephemeris generation PR.

Reply:
Changed here: https://github.com/B612-Asteroid-Institute/adam_core/commit/a580b19ba79f7a9fb27f5cc6e66f96eeff721350 

2-body ephemeris PR isn't ready yet but at least the fix is there.

Reply:
Actually, `not np.all(np.isnan(cov))` is what I was trying to do. I'm trying to check if any values are not NaN. If any values are not NaN, then lets proceed and do the covariance transformation. If all values are NaN, then the covariance transformation will not yield anything useful so lets skip it.

Reply:
Proposed solution: #30